### ISSUE TITLE
Improve Preferences documentation

### ISSUE THREAD

**Issue by fipro78** [eclipse-theia/theia#11497]
### Bug Description:

I tried to add some custom preferences to a Theia application by following the documentation:
https://theia-ide.org/docs/preferences/

I was not able to follow the description to make a contribution work. It is not really an example you can follow. In the end I had to search through the Theia repository and find some preference contributions in the core extensions, e.g.

https://github.com/eclipse-theia/theia/blob/master/packages/editor/src/browser/editor-preferences.ts
https://github.com/eclipse-theia/theia/blob/master/packages/workspace/src/browser/workspace-preferences.ts
https://github.com/eclipse-theia/theia/blob/master/packages/filesystem/src/browser/filesystem-preferences.ts

Looking into these files it looks like they all follow the same pattern:
- Define a preference schema
- Define a configuration interface
- Define some constants and a PreferenceProxy type
- Bind the preferences

This is not covered in the documentation from my point of view. And it seems to be also not current anymore or not complete. At least if I compare the documentation with the mentioned files. The following questions are not answered:

1. Do I need to create my own PreferenceSchema or should I simply use the Theia one? Does it even make any sense to have your own PreferenceSchema?
2. How does a preference value need to be configured? Which properties are available? What types are supported (e.g. is a password field supported)?
3. What is the configuration interface? What is it used/needed for?
4. What are those constants (Symbol?) and the PreferenceProxy? 
5. What is the binding actually doing? I see three bind operations: toDynamicValue, toConstantValue, toService. What are they needed for? What are they actually doing? Does this change the way you can consume preferences?

It should also be mentioned that the bind method in the preferences file needs to be called in the ContainerModule. Probably obvious for someone that is familiar with the whole Theia development process, but for newcomers it is a missing part that should be mentioned.

After reading further to _Listening for a preference change via a configuration_ the last 3 questions get at least a bit explained. But to be honest, I did not read the part about listening for changes as I did not understand completely how to create new preferences.

Maybe the usage of PreferenceProxyFactory should be also added, as this seems to be the recommended way since 1.23, isn't it?

### Additional Information

For newcomers it could be better understandable if you start with a complete class that provides some simple preferences with all recommended configurations. As an example you could use a preference class like this:

```
import { interfaces } from '@theia/core/shared/inversify';
import {
    PreferenceProxy,
    PreferenceSchema,
    PreferenceContribution
} from '@theia/core/lib/browser/preferences';
import { PreferenceProxyFactory } from '@theia/core/lib/browser/preferences/injectable-preference-proxy';

export const dbPreferenceSchema: PreferenceSchema = {
    type: 'object',
    properties: {
        'db.connectionURL': {
            description: 'The URL of the database.',
            type: 'string',
            default: ''
        },
        'db.userName': {
            description: 'The username needed for the database connection.',
            type: 'string',
            default: ''
        },
        'db.userPassword': {
            description: 'The password for the database connection.',
            type: 'string',
            default: ''
        }
    }
};

export interface DBConfiguration {
    'db.connectionURL': String,
    'db.userName': String,
    'db.userPassword': String
}

export const DBPreferenceContribution = Symbol('DBPreferenceContribution');
export const DBPreferences = Symbol('DBPreferences');
export type DBPreferences = PreferenceProxy<DBConfiguration>;

export function bindDBPreferences(bind: interfaces.Bind): void {
    bind(DBPreferences).toDynamicValue(ctx => {
        const factory = ctx.container.get<PreferenceProxyFactory>(PreferenceProxyFactory);
        return factory(dbPreferenceSchema);
    }).inSingletonScope();
    bind(DBPreferenceContribution).toConstantValue({ schema: dbPreferenceSchema });
    bind(PreferenceContribution).toService(DBPreferenceContribution);
}
```
Then explain every section of that class in the documentation step by step, with a link to the section where it is used later (e.g. consuming the preferences as service)

Mention that the bindDBPreferences() needs to be called in the ContainerModule.

Then show how to consume preferences (via PreferenceService or via the injected preferences object):

    @inject(PreferenceService) 
    protected readonly prefService: PreferenceService;

    const url = this.prefService.get('db.connectionURL');

    @inject(Neo4jPreferences)
    protected readonly preferences: Neo4jPreferences;

    const url = this.preferences['db.connectionURL'];

Then add the section about listening for preference changes.

This IMHO is how a user would try to implement custom preferences:
- create the preference contribution
- consume the preferences
- listen for changes

- Operating System: Windows / Linux Remote Container
- Theia Version: 1.28.0
- Documentation Version: Can't tell if that has a version and if it is anyway consistent with the code

I hope this helps in improving the documentation.

**Comment by colin-grant-work** [comment]
I agree that our documentation could use an update, but I think one of your queries is worth addressing now:
> How does a preference value need to be configured? Which properties are available? What types are supported (e.g. is a password field supported)?

Preference schemas are [JSON schemas](https://json-schema.org/), so only the [types](https://json-schema.org/understanding-json-schema/reference/index.html) supported by JSON schemas are available. Those are basically JSON types with a few bits of extra metadata. In particular, no `password` type is supported, and all data stored as preferences will be available in plain text in  a `settings.json` file somewhere in the filesystem, so in general storing passwords as preferences is not a good idea.
